File: arrays/medium/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {

        sort(nums.begin(),nums.end());
        int n = nums.size();
        int target = 0;
        vector<vector<int>> result;
        for(int i=0;i<=n-3;i++){
            if(i>0 && nums[i] == nums[i-1]){
                continue;
            }
            int current_sum = nums[i];
            int p1 = i+1;
            int p2 = n-1;
            while(p1 < p2){
                int rest_sum = nums[p1]+nums[p2];
                if(current_sum+rest_sum == target){
                    result.push_back({current_sum,nums[p1],nums[p2]});
                    p1++;
                    p2--;
                    while(p1 < p2 && nums[p1]==nums[p1-1]){
                        p1++;
                    }
                    while(p1 < p2 && nums[p2]==nums[p2+1]){
                        p2--;
                    }
                }else if(current_sum+rest_sum>target){
                    p2--;
                }else{
                    p1++;
                }
            }
        }
        return result;

    }
};

File: arrays/medium/3sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:

        sorted_nums = sorted(nums)
        results = []
        for i in range(len(nums) - 2):
            if i > 0 and sorted_nums[i] == sorted_nums[i - 1]:
                continue
            current_num = sorted_nums[i]
            diff = 0 - current_num
            p1 = i + 1
            p2 = len(nums) - 1
            while p1 < p2:
                r1 = sorted_nums[p1]
                r2 = sorted_nums[p2]
                if diff == r1 + r2:
                    results.append([current_num, r1, r2])
                    p1 += 1
                    p2 -= 1
                    while p1 < p2 and sorted_nums[p1] == sorted_nums[p1 - 1]:
                        p1 += 1
                    while p1 < p2 and sorted_nums[p2] == sorted_nums[p2 + 1]:
                        p2 -= 1
                elif r1 + r2 < diff:
                    p1 += 1
                else:
                    p2 -= 1
        return results